src/test/java/co/rsk/bitcoinj/script/ScriptTest.java

```diff
@@ -472,84 +468,175 @@ public void numberBuilderNegative() {
     }
 
     @Test
     public void getSigInsertionIndex_fast_bridge_redeem_script() {
         byte[] data = Sha256Hash.of(new byte[]{1}).getBytes();
         Script fastBridgeRedeemScript = RedeemScriptUtils.createFastBridgeRedeemScript(
-            data, btcECKeyList);
+            data, federationKeys);
 
         testGetSigInsertionIndex(fastBridgeRedeemScript);
     }
 
     @Test
     public void getSigInsertionIndex_erp_redeem_script() {
         Script erpRedeemScript = RedeemScriptUtils.createErpRedeemScript(
-            btcECKeyList,
-            erpFedECKeyList,
+            federationKeys,
+            erpFederationKeys,
             500L
         );
 
         testGetSigInsertionIndex(erpRedeemScript);
     }
 
+    @Test
+    public void getSigInsertionIndex_whenP2shRedeemScript_shouldReturnIndex() {
+        Script redeemScript = RedeemScriptUtils.createP2shErpRedeemScript(
+            federationKeys,
+            erpFederationKeys,
+            600L
+        );
+
+        testGetSigInsertionIndex(redeemScript);
+    }
+
     @Test
     public void getSigInsertionIndex_fast_bridge_erp_redeem_script() {
         Script fastBridgeErpRedeemScript = RedeemScriptUtils.createFastBridgeErpRedeemScript(
-            btcECKeyList,
-            erpFedECKeyList,
+            federationKeys,
+            erpFederationKeys,
             500L,
             Sha256Hash.of(new byte[]{1}).getBytes()
         );
 
         testGetSigInsertionIndex(fastBridgeErpRedeemScript);
     }
 
+    @Test
+    public void getSigInsertionIndex_whenFlyoverP2shRedeemScript_shouldReturnIndex() {
+        Script flyoverP2shRedeemScript = RedeemScriptUtils.createFastBridgeP2shErpRedeemScript(
+            federationKeys,
+            erpFederationKeys,
+            500L,
+            Sha256Hash.of(new byte[]{1}).getBytes()
+        );
+
+        testGetSigInsertionIndex(flyoverP2shRedeemScript);
+    }
+
     @Test
     public void getSigInsertionIndex_no_fast_bridge_redeem_script() {
-        Script redeemScript = RedeemScriptUtils.createStandardRedeemScript(btcECKeyList);
+        Script redeemScript = RedeemScriptUtils.createStandardRedeemScript(federationKeys);
         testGetSigInsertionIndex(redeemScript);
     }
 
+    @Test
+    public void getSigInsertionIndex_whenEmptyScript_shouldReturnZero() {
+        Script redeemScript = new ScriptBuilder().build();
+        Sha256Hash hashForSignature = Sha256Hash.of(new byte[]{1});
+        BtcECKey signingKey = BtcECKey.fromPrivate(BigInteger.valueOf(800));
+
+        int sigInsertionIndex = redeemScript.getSigInsertionIndex(hashForSignature, signingKey);
+        Assert.assertEquals(expectedDefaultSigInsertionIndexForNoRedeemScript, sigInsertionIndex);
+
+        assertScriptIsNoRedeemScript(redeemScript, sigInsertionIndex, hashForSignature, signingKey);
+    }
+
+    @Test
+    public void getSigInsertionIndex_whenP2shScript_shouldReturnZero() {
```

Comment:
Why return 0 for a p2sh script? Isn't that a valid redeem script?

Reply:
It isn't. The `RedeemScriptParserFactory` identified the p2shScriptOutputs as `NoRedeemScriptParser`.